### PR TITLE
Mission download

### PR DESCRIPTION
**Pull request by julianoes**
This adds mission download and makes sure we compare the mission items uploaded and downloaded again.

### CODE REVIEW

File: src/tests/base.h [hunk 1]
```diff
@@ -7,6 +7,8 @@
 #include <memory>
 #include <string>
 
+#define UNUSED(x) (void)(x)
```
Comment (bkueng) [inline src/tests/base.h:15]: I'd rather disable the warning

File: .clang-format [hunk 2]
```diff
@@ -1,6 +1,6 @@
 Language: Cpp
 BasedOnStyle: Google
-ColumnLimit: 120
+ColumnLimit: 100
```
Comment (bkueng) [inline .clang-format:3]: I'm fine with it but:
- I don't want to have more-or-less arbitrary style changes based on personal preferences.
- This should preferably be a separate PR.

**Review by bkueng — COMMENTED** [review]
Looks good, thanks! One thing: can you remove the need for the `UNUSED` macro by disabling the warning instead?

File: .clang-format (hunk 2 above)
Comment (julianoes) [reply .clang-format:3]: Okay!

File: src/tests/base.h (hunk 1 above)
Comment (julianoes) [reply src/tests/base.h:15]: Why? This makes it explicit.

Comment (bkueng) [reply src/tests/base.h:15]: I don't see why it needs to be explicit, I have not come across a case where an argument was accidentally ignored (not saying it doesn't happen though). Ignoring it due to inheritance or for other reasons is however quite common.

Comment (julianoes) [reply src/tests/base.h:15]: :thinking: right.

File: src/tests/mission.h [hunk 3]
```diff
@@ -5,13 +5,16 @@
 #include <dronecode_sdk/dronecode_sdk.h>
 #include <dronecode_sdk/plugins/mission/mission.h>
 
+namespace dcsdk = dronecode_sdk;
+using mission_items = std::vector<std::shared_ptr<dcsdk::MissionItem>>;
```
Comment (bkueng) [inline src/tests/mission.h]: Can you move this into the class?

File: src/tests/base.cpp [hunk 4]
```diff
@@ -19,6 +22,28 @@ void TestBase::storeConfig(ConfigNode& config)
 
 TestBase::TestBase(const Context& context) : _context(context) {}
 
+std::string TestBase::extractFilename(const std::string& path)
+{
+    std::stringstream ss(path);
+    std::string token;
+    while (std::getline(ss, token, '/'))
```
Comment (bkueng) [inline src/tests/base.cpp]: Why not using `std::string::find_last_of`?

File: src/tests/base.h (hunk 1 above)
Comment (bkueng) [reply src/tests/base.h:15]: So what's your take on this?

File: src/tests/base.h [hunk 5]
```diff
@@ -3,9 +3,45 @@
...
+#include <iostream>
 #include <map>
 #include <memory>
+#include <sstream>
 #include <string>
+#include <type_traits>
+
+#define UNUSED(x) (void)(x)
+
+// SFINAE to check if something is streamable adapted from:
```
Comment (bkueng) [inline src/tests/base.h]: Can you move this in some other helper/util header?

File: src/tests/base.h [hunk 6]
```diff
@@ -62,9 +98,37 @@ class TestBase
...
+        if (a == b) {
+            return;
+        }
+
+        std::cout << "Check at " << extractFilename(file) << ":" << line << " failed" << std::endl;
+        std::cout << a_str << " != " << b_str << std::endl;
+
+        auto maybe_streamable = MaybeStreamable<T>();
+
+        if (maybe_streamable.isStreamable()) {
```
Comment (bkueng) [inline src/tests/base.h:104]: Nice

File: src/tests/base.h [hunk 7]
```diff
@@ -62,9 +98,37 @@ class TestBase
...
+            std::cout << maybe_streamable.print(a) << " != " << maybe_streamable.print(b)
+                      << std::endl;
+        }
+
+        _result = Result::Failed;
```
Comment (bkueng) [inline src/tests/base.h:109]: I'd rather throw an exception here. Continue running a failed test does not make much sense to me.

**Review by bkueng — COMMENTED** [review]
I like the `EXPECT_EQ`, it simplifies the code quite a bit.

File: src/tests/base.h (hunk 5 above)
Comment (julianoes) [reply src/tests/base.h]: Ok.

File: src/tests/base.cpp (hunk 4 above)
Comment (julianoes) [reply src/tests/base.cpp]: That seems easier!

File: src/tests/base.h (hunk 7 above)
Comment (julianoes) [reply src/tests/base.h:109]: It depends on the check. That's why usually there are `expects` and `asserts`. expects should just take note, and asserts should stop.

Comment (bkueng) [reply src/tests/base.h:109]: Fine, then we also need an assert.

Comment (julianoes) [reply src/tests/base.h:109]: Ok.

File: src/tests/mission.h (hunk 3 above)
Comment (julianoes) [reply src/tests/mission.h]: I removed it.

File: src/tests/base.cpp (hunk 4 above)
Comment (julianoes) [reply src/tests/base.cpp]: Implemented, thx!

File: src/tests/base.cpp [hunk 8]
```diff
@@ -19,6 +22,25 @@ void TestBase::storeConfig(ConfigNode& config)
 
 TestBase::TestBase(const Context& context) : _context(context) {}
 
+std::string TestBase::extractFilename(const std::string& path)
+{
+    int pos = path.find_last_of('/');
```
Comment (bkueng) [inline src/tests/base.cpp]: To be on the safe side, can you handle `string::npos` as return value?

File: src/tests/base.h (hunk 5 above)
Comment (julianoes) [reply src/tests/base.h]: Moved.

File: src/tests/base.h (hunk 7 above)
Comment (julianoes) [reply src/tests/base.h:109]: Implemented.

File: src/tests/base.cpp (hunk 8 above)
Comment (julianoes) [reply src/tests/base.cpp]: Will do.

**Comment by julianoes** [comment]
And now we can merge this one and then I'll move to other things.

File: src/tests/base.cpp [hunk 9]
```diff
@@ -19,6 +22,29 @@ void TestBase::storeConfig(ConfigNode& config)
 
 TestBase::TestBase(const Context& context) : _context(context) {}
 
+std::string TestBase::extractFilename(const std::string& path)
+{
+    auto pos = path.find_last_of('/');
```
Comment (bkueng) [inline src/tests/base.cpp]: I guess we should handle windows as well: '\\\\'

File: src/tests/mission.cpp [hunk 10]
```diff
@@ -6,49 +6,91 @@
...
 
     // wait until uploaded
-    const Mission::Result result = future_result.get();
-    if (result != Mission::Result::SUCCESS) {
-        cout << "Mission upload failed (" << Mission::result_str(result) << "), exiting." << endl;
-        return Result::Failed;
-    }
-    cout << "Mission uploaded." << endl;
+    const dcsdk::Mission::Result result = fut.get();
+    EXPECT_EQ(result, dcsdk::Mission::Result::SUCCESS);
```
Comment (bkueng) [inline src/tests/mission.cpp]: If the upload fails, there's probably no point in downloading it. Meaning this should be an assertion.

File: src/tests/mission.cpp [hunk 11]
```diff
@@ -6,49 +6,91 @@
...
+    // wait until uploaded
+    EXPECT_EQ(result, dcsdk::Mission::Result::SUCCESS);
+
+    return items;
+}
+
+void Mission::compareMissions(const std::vector<std::shared_ptr<dcsdk::MissionItem>>& items_a,
+                              const std::vector<std::shared_ptr<dcsdk::MissionItem>>& items_b)
+{
+    EXPECT_EQ(items_a.size(), items_b.size());
```
Comment (bkueng) [inline src/tests/mission.cpp:89]: This needs to be an assertion. Otherwise you can access invalid data.

File: src/tests/param.cpp [hunk 12]
```diff
@@ -14,43 +14,27 @@ REGISTER_TEST(ParamChange);
...
-    }
+    EXPECT_EQ(get_result1.first, ParamsRaw::Result::SUCCESS);
+    EXPECT_EQ(get_result1.second, _config.set_value);
 
     cout << "Setting param " << _config.name << " to " << _config.reset_value << endl;
     ParamsRaw::Result set_result2 = _params_raw.set_param_int(_config.name, _config.reset_value);
-    if (set_result2 != ParamsRaw::Result::SUCCESS) {
-        return Result::Failed;
-    }
+    EXPECT_EQ(set_result2, ParamsRaw::Result::SUCCESS);
```
Comment (bkueng) [inline src/tests/param.cpp:31]: Similar here: if setting the parameter fails, it's very likely that a read will fail as well (at least the value comparison).

**Review by bkueng — COMMENTED** [review]
I think most of the checks should be assertions.
I'm also wondering if the asserted exception should be caught, and continue with the next test, instead of aborting everything.

File: src/tests/base.cpp (hunk 9 above)
Comment (julianoes) [reply src/tests/base.cpp]: Fixed in 6af65e4.

File: src/tests/mission.cpp (hunk 10 above)
Comment (julianoes) [reply src/tests/mission.cpp]: It might have been uploaded anyway but we got a timeout. I think we can still try to download it, and most probably it'll be wrong.

File: src/tests/mission.cpp (hunk 11 above)
Comment (julianoes) [reply src/tests/mission.cpp:89]: True.

File: src/tests/param.cpp (hunk 12 above)
Comment (julianoes) [reply src/tests/param.cpp:31]: Which is fine I think.

File: src/tests/mission.cpp (hunk 11 above)
Comment (julianoes) [reply src/tests/mission.cpp:89]: Fixed in d27d0f8.

File: src/tests/param.cpp (hunk 12 above)
Comment (bkueng) [reply src/tests/param.cpp:31]: It's a question of what we want. It can be quite confusing, since you will see failures that are a consequence of other failures.

Comment (julianoes) [reply src/tests/param.cpp:31]: Right, so maybe we should use asserts more often as you suggest but then have them per test (e.g. Mission) only.

Comment (bkueng) [reply src/tests/param.cpp:31]: Exactly, that's what I was thinking.

**Review by bkueng — COMMENTED** [review]
Not sure if you haven't gotten around to do so yet or forgot: can you have another pass at the code and replace EXPECT_EQ with ASSERT_EQ where needed? It's good to merge afterwards.

File: src/tests/param.cpp (hunk 12 above)
Comment (julianoes) [reply src/tests/param.cpp:31]: Addressed in 2f8054d05e1fb33086702d51f749a1dd84ae67c5.

**Comment by julianoes** [comment]
> Not sure if you haven't gotten around to do so yet or forgot: can you have another pass at the code and replace EXPECT_EQ with ASSERT_EQ where needed? It's good to merge afterwards.

Done, I think.

**Review by bkueng — APPROVED** [review]
